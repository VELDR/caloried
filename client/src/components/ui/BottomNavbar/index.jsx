import { useLocation, useNavigate } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { AutoStories, Dashboard, Fastfood, Person } from '@mui/icons-material';
import SearchBar from '@components/ui/SearchBar';

import classes from './style.module.scss';

const BottomNavbar = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const isActive = (path) => location.pathname === path;

  const getItemClassName = (path) => `${classes.navbar__item} ${isActive(path) ? classes.active : ''}`;

  return (
    <div className={classes.bottomNav} data-testid="bottom-navbar">
      <div className={classes.searchBar}>
        <SearchBar />
      </div>
      <div className={classes.navbar}>
        <div
          className={getItemClassName('/dashboard')}
          onClick={() => navigate('/dashboard')}
          data-testid="bottom-navigate-dashboard"
        >
          <Dashboard />
          <div className={classes.label}>
            <FormattedMessage id="app_dashboard" />
          </div>
        </div>
        <div
          className={getItemClassName('/diary')}
          onClick={() => navigate('/diary')}
          data-testid="bottom-navigate-diary"
        >
          <AutoStories />
          <div className={classes.label}>
            <FormattedMessage id="app_diary" />
          </div>
        </div>
        <div
          className={getItemClassName('/search')}
          onClick={() => navigate('/search')}
          data-testid="bottom-navigate-search"
        >
          <Fastfood />
          <div className={classes.label}>
            <FormattedMessage id="app_food" />
          </div>
        </div>
        <div
          className={getItemClassName('/profile')}
          onClick={() => navigate('/profile')}
          data-testid="bottom-navigate-profile"
        >
          <Person />
          <div className={classes.label}>
            <FormattedMessage id="app_profile" />
          </div>
        </div>
      </div>
    </div>
  );
};

export default BottomNavbar;
