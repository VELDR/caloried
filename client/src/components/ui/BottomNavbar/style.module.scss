@import '@styles/base';

.bottomNav {
  position: fixed;
  bottom: 0;
  right: 0;
  left: 0;
  width: 100%;
  display: block;
  @include flex-column(0.5rem);
  z-index: 100;
  @include apply-at-sm-size {
    display: none;
  }
  .searchBar {
    padding: 0 0.5rem;
  }
  .navbar {
    @include flex-items-center;
    flex-wrap: nowrap;
    justify-content: space-between;
    flex-direction: row;
    width: 100%;
    height: 4rem;
    background-color: var(--color-bg-primary);
    border-top: 1px solid var(--color-text-secondary);
    &__item {
      @include flex-center-column;
      font-size: $font-size-small;
      width: 100%;
      color: var(--color-text-primary);
      padding: 0 1rem;
      overflow: hidden;
      cursor: pointer;
      transition:
        width 200ms,
        padding 200ms;
      transition-timing-function: ease-in-out;
      &:hover {
        background-position: 0 50%;
        padding: 1rem;
        svg {
          font-size: $font-size-larger;
        }
        .label {
          opacity: 1;
        }
      }
      svg {
        font-size: $font-size-huge;
        position: relative;
        z-index: 2;
        transition: font-size 250ms;
        transition-timing-function: ease-out;
      }
      .label {
        font-weight: 600;
        letter-spacing: 0.15em;
        opacity: 0;
        position: relative;
        z-index: 2;
        font-size: $font-size-smaller;
        transition: opacity 400ms;
        transition-delay: 130ms;
        transition-timing-function: ease-out;
      }
    }
    .active {
      svg {
        color: var(--color-primary);
      }
      .label {
        color: var(--color-primary);
        opacity: 1;
      }
    }
  }
}
