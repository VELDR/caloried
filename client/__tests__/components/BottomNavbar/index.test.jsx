import BottomNavbar from '@components/ui/BottomNavbar';
import { fireEvent, render } from '@utils/testHelper';

const mockNavigate = jest.fn();
const mockLocation = {
  pathname: '/dashboard',
};

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
  useLocation: () => mockLocation,
}));

let wrapper;

beforeEach(() => {
  wrapper = render(<BottomNavbar />);
});

describe('BottomNavbar Component', () => {
  it('should render correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('bottom-navbar')).toBeInTheDocument();
  });
  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should call navigate to dashboard when dashboard button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('bottom-navigate-dashboard');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/dashboard`);
  });
  it('should call navigate to diary when diary button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('bottom-navigate-diary');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/diary`);
  });
  it('should call navigate to search when food button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('bottom-navigate-search');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/search`);
  });
  it('should call navigate to profile when profile button is clicked', () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('bottom-navigate-profile');
    fireEvent.click(button);
    expect(mockNavigate).toHaveBeenCalledWith(`/profile`);
  });
  it('should apply the correct classes for active and inactive buttons', () => {
    const { getByTestId } = wrapper;

    const dashboardButton = getByTestId('bottom-navigate-dashboard');
    expect(dashboardButton).toHaveClass('active');
    expect(dashboardButton).toHaveClass('navbar__item');

    const diaryButton = getByTestId('bottom-navigate-diary');
    expect(diaryButton).not.toHaveClass('active');
    expect(diaryButton).toHaveClass('navbar__item');

    mockLocation.pathname = '/diary';
    wrapper.rerender(<BottomNavbar />);

    expect(diaryButton).toHaveClass('active');
  });
});
