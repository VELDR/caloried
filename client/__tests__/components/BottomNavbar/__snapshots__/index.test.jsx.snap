// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomNavbar Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bottomNav"
        data-testid="bottom-navbar"
      >
        <div
          class="searchBar"
        >
          <div
            class="search"
            data-testid="search-bar"
          >
            <input
              class="search__input"
              placeholder="Cari Makanan"
              type="search"
              value=""
            />
            <div
              class="icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="navbar"
        >
          <div
            class="navbar__item active"
            data-testid="bottom-navigate-dashboard"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DashboardIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
              />
            </svg>
            <div
              class="label"
            >
              Dashbor
            </div>
          </div>
          <div
            class="navbar__item "
            data-testid="bottom-navigate-diary"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AutoStoriesIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m19 1-5 5v11l5-4.5V1zM1 6v14.65c0 .25.25.5.5.5.1 0 .15-.05.25-.05C3.1 20.45 5.05 20 6.5 20c1.95 0 4.05.4 5.5 1.5V6c-1.45-1.1-3.55-1.5-5.5-1.5S2.45 4.9 1 6zm22 13.5V6c-.6-.45-1.25-.75-2-1v13.5c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5v2c1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.3 4.75 1.05.1.05.15.05.25.05.25 0 .5-.25.5-.5v-1.1z"
              />
            </svg>
            <div
              class="label"
            >
              Diary
            </div>
          </div>
          <div
            class="navbar__item "
            data-testid="bottom-navigate-search"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FastfoodIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.06 22.99h1.66c.84 0 1.53-.64 1.63-1.46L23 5.05h-5V1h-1.97v4.05h-4.97l.3 2.34c1.71.47 3.31 1.32 4.27 2.26 1.44 1.42 2.43 2.89 2.43 5.29v8.05zM1 21.99V21h15.03v.99c0 .55-.45 1-1.01 1H2.01c-.56 0-1.01-.45-1.01-1zm15.03-7c0-8-15.03-8-15.03 0h15.03zM1.02 17h15v2h-15z"
              />
            </svg>
            <div
              class="label"
            >
              Makanan
            </div>
          </div>
          <div
            class="navbar__item "
            data-testid="bottom-navigate-profile"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
            <div
              class="label"
            >
              Profil
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bottomNav"
      data-testid="bottom-navbar"
    >
      <div
        class="searchBar"
      >
        <div
          class="search"
          data-testid="search-bar"
        >
          <input
            class="search__input"
            placeholder="Cari Makanan"
            type="search"
            value=""
          />
          <div
            class="icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="navbar"
      >
        <div
          class="navbar__item active"
          data-testid="bottom-navigate-dashboard"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DashboardIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 13h8V3H3v10zm0 8h8v-6H3v6zm10 0h8V11h-8v10zm0-18v6h8V3h-8z"
            />
          </svg>
          <div
            class="label"
          >
            Dashbor
          </div>
        </div>
        <div
          class="navbar__item "
          data-testid="bottom-navigate-diary"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AutoStoriesIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m19 1-5 5v11l5-4.5V1zM1 6v14.65c0 .25.25.5.5.5.1 0 .15-.05.25-.05C3.1 20.45 5.05 20 6.5 20c1.95 0 4.05.4 5.5 1.5V6c-1.45-1.1-3.55-1.5-5.5-1.5S2.45 4.9 1 6zm22 13.5V6c-.6-.45-1.25-.75-2-1v13.5c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5v2c1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.3 4.75 1.05.1.05.15.05.25.05.25 0 .5-.25.5-.5v-1.1z"
            />
          </svg>
          <div
            class="label"
          >
            Diary
          </div>
        </div>
        <div
          class="navbar__item "
          data-testid="bottom-navigate-search"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FastfoodIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18.06 22.99h1.66c.84 0 1.53-.64 1.63-1.46L23 5.05h-5V1h-1.97v4.05h-4.97l.3 2.34c1.71.47 3.31 1.32 4.27 2.26 1.44 1.42 2.43 2.89 2.43 5.29v8.05zM1 21.99V21h15.03v.99c0 .55-.45 1-1.01 1H2.01c-.56 0-1.01-.45-1.01-1zm15.03-7c0-8-15.03-8-15.03 0h15.03zM1.02 17h15v2h-15z"
            />
          </svg>
          <div
            class="label"
          >
            Makanan
          </div>
        </div>
        <div
          class="navbar__item "
          data-testid="bottom-navigate-profile"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
          <div
            class="label"
          >
            Profil
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
